app/models/name/parse.rb

```diff
@@ -207,7 +216,7 @@ def self.parse_name(str, rank: :Genus, deprecated: false)
   # Guess rank of +text_name+.
   def self.guess_rank(text_name)
     TEXT_NAME_MATCHERS.each do |matcher|
-      return matcher.second if matcher.first =~ text_name
+      return matcher.second if text_name.match?(matcher.first)
```

Comment:
This (and lines 356-70) below were were my response to RuboCop's [Style/MultilineTernaryOperator](https://rubocop.readthedocs.io/en/latest/cops_style/#stylemultilineternaryoperator) and [Style/NestedTernaryOperator](https://rubocop.readthedocs.io/en/latest/cops_style/#stylenestedternaryoperator).
But my solution (a dispatch table based on an array) strikes me as ugly and inefficient.
Is there a better way to do this? (Something more OO/Ruby/Rails-like?)

Reply:
I don't understand your concern yet.  When I compare this file to what is currently on master, I don't see any significant changes to ternary operators.  A change I would at least consider is using dictionary rather than an array.  Since Ruby 1.9 hashes are guaranteed to preserve order.  It would seem clearer to me if the ranks were the keys, but then the values would need to be arrays of possible values and you'd end having to go through those lists.  Given that I'd also consider creating a small light-weight that has a regex and rank and can do the match function.  It would be tiny bit clearer, but it wouldn't really be different from what you have here.

Reply:
1. The change to deal with NestedTernaryOperator was #444. It used to read:
```ruby
  def self.guess_rank(text_name)
    / (group|clade|complex)$/.match?(text_name) ? :Group :
    text_name.include?(" f. ")         ? :Form       :
    text_name.include?(" var. ")       ? :Variety    :
    text_name.include?(" subsp. ")     ? :Subspecies :
    text_name.include?(" stirps ")     ? :Stirps     :
    text_name.include?(" subsect. ")   ? :Subsection :
    text_name.include?(" sect. ")      ? :Section    :
    text_name.include?(" subgenus ")   ? :Subgenus   :
    text_name.include?(" ")            ? :Species    :
    /^\S+aceae$/.match?(text_name)     ? :Family     :
    /^\S+ineae$/.match?(text_name)     ? :Family     : # :Suborder
    /^\S+ales$/.match?(text_name)      ? :Order      :
    /^\S+mycetidae$/.match?(text_name) ? :Order      : # :Subclass
    /^\S+mycetes$/.match?(text_name)   ? :Class      :
    /^\S+mycotina$/.match?(text_name)  ? :Class      : # :Subphylum
    /^\S+mycota$/.match?(text_name)    ? :Phylum     :
    /^Fossil-/.match?(text_name)       ? :Phylum     :
                                         :Genus
  end
```
2. There's something missing in one sentence of your comment: "'... I'd also consider creating a small light-weight that has a regex and rank and can do the match function." I assume that means "a small light-weight **class**". (I just want to be sure about any new classes, as my mind is stuck on procedural programming and still has trouble dealing with OOP.)

Reply:
Yes.  I meant class.  That original code is kind of scary!  Either an array of arrays or a dictionary is better.

Reply:
I haven't tested it, but something like:
```
class RankMatcher
  attr_reader :rank
  attr_reader :pattern

  def initialize(rank, pattern)
    @rank = rank
    @pattern = pattern
  end

  def match?(text_name)
    return text_name.match?(@pattern)
  end
end

TEXT_NAME_MATCHERS = [
  RankMatcher(:Group, / (group|clade|complex)$/),
  ...
  RankMatcher(:Genus, //)]

def self.guess_rank(text_name)
  return TEXT_NAME_MATCHERS.find{ |matcher| matcher.match?(text_name) }
end
```

Reply:
Thanks very much for the response and especially for the example code. 
I will try implementing it, but probably won't start until tomorrow p.m.

Reply:
It would also be slightly safer/more explicit to remove `RankMatcher(:Genus, //)` from TEXT_NAME_MATCHERS and add it as the `ifnone` argument to `find`.

Reply:
Thanks again! I had been wondering about that.

Reply:
If you're having fun with this, then it would be even better to move the lists into the same file as the light-weight class and just have a TEXT_NAME_RANK_MATCHER and ABBREV_RANK_MATCHER defined in the class file that which parse.rb would then use.  This would imply moving a bunch of the other constants our of parse.rb and might suggest some other ways to break it up all of which would be a good thing, but potentially quite a bit more work.

Reply:
I am indeed enjoying this. I'll probably have time to move the two listed matchers, not time for the rest of the constants. (I'll email you & @pellaea about my schedule.)
What do you think about keeping the lightweight class, with it's moved constants, in the same file? My thought was it might be easier to understand both parse.rb and the lightweight class if one could read them together.

Reply:
Personally I lean towards one class per file especially given how large the files have become.  For classes that are only used by one other class or are derived from a common base class, I sometimes create sub-folders for these classes, but only if there are multiple such classes and ideally when their are for similar cases.  The example of a TextNameRankMatcher and an AbbrevRankMatcher class would fit this case nicely.  In my view, a NameParser class would be better than stuffing all this code into the Name model.